Streaming bulk operations

chrismwendt opened the pull request:
This adds streaming bulk functionality via Ruby's Enumerators.

This does not to alter any of the existing methods in backwards incompatible ways so that a conversation about making the API more consistent could take place later. Because of that, some of the small operation methods had to be nearly duplicated.

This solves #92.

/cc @TwP @grantr

chrismwendt commented on lib/elastomer/client/bulk.rb:
```diff
@@ -43,6 +43,48 @@ def bulk( body = nil, params = nil )
+          end
+        end
+      end
+    end
```
Is there a simpler way to essentially `map` over an `Enumerator` than by constructing a new one and yielding each element of the original, as I did here?

TwP replied:
On thought on this method is to **not** return an `Enumerator`. Instead, accept a block and perform the iteration directly with this method.

```ruby
def bulk_stream_items(ops, params = {})
  stats = {
    "took" => 0,
    "errors" => false,
    "success" => 0,
    "failure" => 0
  }

  bulk_stream_responses(ops, params).each do |response|
    stats["took"] += response["took"]
    stats["errors"] |= response["errors"]

    response["items"].each do |item|
      if is_ok?(item)
        stats["success"] += 1
      else
        stats["failure"] += 1
      end
      yield item
    end
  end

  return stats
end
```

I've also added a hash of interesting numbers as the return value from this method. The `is_ok?` method does not exist yet, but it would check the HTTP status returned with each item.

```
def is_ok?(item)
  item["status"].between?(200, 299)
end
```

Thoughts on this approach?

chrismwendt replied:
The client might want to retry failed documents, in which case they would need access to at least the failed response items. Maybe it's worth having a separate convenient method called `bulk_stream_stats` which does exactly what you suggest.

TwP replied:
The client does have access to the failed items during the iteration. When `yield item` is called, the client can also check the status code of that particular operation.

Returning the stats hash is there just to be nice to the end user.

chrismwendt replied:
My apologies, I read over your code too quickly and missed that `yield item` :bow:

I like the convenience of returning stats, and the block syntax would look almost the same as calling `each` on the `Enumerator`. I'll implement it like you showed.

TwP commented on lib/elastomer/client/bulk.rb:
```diff
@@ -43,6 +43,48 @@ def bulk( body = nil, params = nil )
+    #     Bulk::index(document1, :_type => 'default-type', :_id => 1),
+    #     Bulk::index(document2, params2),
+    #     Bulk::delete(params3),
+    #   ]
```
I'm not a huge fan of the `Bulk::index` / `Bulk::delete` syntax here. My proposal is this:

```ruby
ops = [
  [:index, document1, {:_type => "foo", :_id => 1}],
  [:create, document2],
  [:delete, {:_type => "bar", :_id => 42}]
]
```

The first item in the list is the bulk action to take. This corresponds to one of the action methods on the Bulk instance. The second item in the list is a document (or the params in the case of `:delete`). And thirdly comes the optional params.

Inside the Enumerator we have:

```ruby
Enumerator.new do |yielder|
  ops.each do |action, *args|
    response = bulk_obj.send(action, *args)
    yielder.yield response unless response.nil?
  end
  response = bulk_obj.call
  yielder.yield response unless response.nil?
end
```

The Ruby `send` method is being used here to call the `action` method on the `bulk_obj` and passing the `*args` to the method.

chrismwendt replied:
That's a pretty clever use of `send` to reuse the existing methods! I'll implement that now.

TwP commented on lib/elastomer/client/bulk.rb:
```diff
@@ -211,7 +295,7 @@ def call
       # Internal: convert special key parameters to their wire representation
       # and apply any override document parameters.
-      def prepare_params(document, params)
+      def self.prepare_params(document, params)
```
This can be reverted back to an instance method.

TwP commented on lib/elastomer/client/bulk.rb:
```diff
@@ -227,7 +311,7 @@ def prepare_params(document, params)
       #
       # Returns extracted key/value pairs as a Hash.
-      def from_document( document )
+      def self.from_document( document )
```
This can be reverted back to an instance method.

TwP commented on lib/elastomer/client/bulk.rb:
```diff
@@ -249,7 +333,7 @@ def from_document( document )
       #
       # Returns a new params Hash with the special keys replaced.
-      def convert_special_keys(params)
+      def self.convert_special_keys(params)
```
This can be reverted back to an instance method.

grantr commented on lib/elastomer/client/bulk.rb:
```diff
@@ -43,6 +43,90 @@ def bulk( body = nil, params = nil )
+    #     [:delete, {:_type => "bar", :_id => 42}]
+    #   ]
+    #   bulk_stream_items(ops, :index => 'default-index') do |item|
+    #     puts item
```
Can we document what the item typically looks like?

chrismwendt commented:
@TwP: I reverted those class methods back to instance methods.

@grantr: I added a sample response item to the docs.

I also bumped the version and updated the changelog.

TwP commented:
:+1: :shipit: :sparkles: and that's enough emoji for now.

grantr commented:
This is great stuff, useful and easy to use. :+1:
